src/python/pants/core/goals/tailor.py

```diff
@@ -0,0 +1,382 @@
+
+    @classmethod
+    def merge(cls, tgts_iters: Iterable[PutativeTargets]) -> PutativeTargets:
+        all_tgts: List[PutativeTarget] = []
+        for tgts in tgts_iters:
+            all_tgts.extend(tgts)
+        return cls(all_tgts)
+
+
+class TailorSubsystem(GoalSubsystem):
+    name = "tailor"
+    help = "Generate config for working with Pants."
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--build-file-indent",
+            advanced=True,
+            type=str,
+            default="    ",
+            help="The indent to use when auto-editing BUILD files.",
+        )
+
+    @property
+    def build_file_indent(self) -> str:
+        return cast(str, self.options.build_file_indent)
+
+
+class Tailor(Goal):
+    subsystem_cls = TailorSubsystem
+
+
+def group_by_build_file(ptgts: Iterable[PutativeTarget]) -> Dict[str, List[PutativeTarget]]:
+    ret = defaultdict(list)
+    for ptgt in ptgts:
+        ret[ptgt.build_file_path].append(ptgt)
+    return ret
+
+
+@dataclass(frozen=True)
+class UniquelyNamedPutativeTargets:
+    """Putative targets that have no name conflicts with existing targets (or each other)."""
+
+    putative_targets: PutativeTargets
+
+
+@rule
+async def rename_conflicting_targets(ptgts: PutativeTargets) -> UniquelyNamedPutativeTargets:
+    """Ensure that no target addresses collide."""
+    all_existing_tgts = await Get(
+        UnexpandedTargets, AddressSpecs([MaybeEmptyDescendantAddresses("")])
+    )
+    existing_addrs: Set[str] = {tgt.address.spec for tgt in all_existing_tgts}
+    uniquely_named_putative_targets: List[PutativeTarget] = []
+    for ptgt in ptgts:
+        idx = 0
+        possibly_renamed_ptgt = ptgt
+        while possibly_renamed_ptgt.address.spec in existing_addrs:
+            possibly_renamed_ptgt = ptgt.rename(f"{ptgt.name}{idx}")
+            idx += 1
+        uniquely_named_putative_targets.append(possibly_renamed_ptgt)
+        existing_addrs.add(possibly_renamed_ptgt.address.spec)
+
+    return UniquelyNamedPutativeTargets(PutativeTargets(uniquely_named_putative_targets))
+
+
+@dataclass(frozen=True)
+class DisjointSourcePutativeTarget:
+    """Putative target whose sources don't overlap with those of any existing targets."""
+
+    putative_target: PutativeTarget
+
+
+@rule
+async def restrict_conflicting_sources(ptgt: PutativeTarget) -> DisjointSourcePutativeTarget:
+    source_paths = await Get(
+        Paths,
+        PathGlobs(Sources.prefix_glob_with_dirpath(ptgt.path, glob) for glob in ptgt.owned_sources),
+    )
+    source_path_set = set(source_paths.files)
+    source_dirs = {os.path.dirname(path) for path in source_path_set}
+    possible_owners = await Get(Targets, AddressSpecs(AscendantAddresses(d) for d in source_dirs))
+    possible_owners_sources = await MultiGet(
+        Get(SourcesPaths, SourcesPathsRequest(t.get(Sources))) for t in possible_owners
+    )
```

Comment:
Note that you can directly request owning targets via: https://github.com/pantsbuild/pants/blob/6903b62d6af605e55dbd202163d1a72435083507/src/python/pants/engine/internals/graph.py#L312-L325

Reply:
I purposely didn't use that because it also finds secondary owners, and does some analysis to compare deleted to live files, and it returns file targets whereas we want BUILD file targets only (although this did make me realize that there's a bug here, and we want to get `UnexpandedTargets` here, so I fixed that).

Reply:
@eric does the latest change look right to you?

Reply:
Yeah, that sounds sensible to not use `Owners`, including to avoid live vs. deleted files.